src/Stratis.Bitcoin/Utilities/AsyncConsumerQueue.cs

```diff
@@ -0,0 +1,154 @@
+        {
+            this.queue = new ConcurrentQueue<T>();
+            this.processingAction = processingAction;
+            this.immediateProcessingTrigger = immediateProcessingTrigger;
+            this.timer = timer;
+            this.restartTimerIfNotAllItemsConsumed = restartTimerIfNotAllItemsConsumed;
+
+            this.trigger = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
+            this.lockObject = new object();
+        }
+
+        /// <inheritdoc />
+        public void Enqueue(T item)
+        {
+            this.queue.Enqueue(item);
+
+            Task.Run(() =>
+            {
+                if (this.immediateProcessingTrigger != null && this.immediateProcessingTrigger(item))
+                {
+                    this.trigger.Set();
+                }
+                else
+                {
+                    lock (this.lockObject)
+                    {
+                        if (this.timer != null && (this.timerAwaiterTask == null))
+                            SetTimer();
+                    }
+                }
+
+                lock (this.lockObject)
+                {
+                    // Ensure flow is running.
+                    if (this.executor == null || this.executor.IsCompleted)
+                        this.executor = ExecutionFlowAsync();
+                }
+
+            }).ConfigureAwait(false);
+        }
+
+        private void SetTimer()
+        {
+            this.timerAwaiterTask = Task.Delay(this.timer.Value, this.cancellation.Token);
+        }
+
+        private async Task ExecutionFlowAsync()
+        {
+            // Create task that waits for the trigger event.
+            if (this.triggerAwaiterTask == null || this.triggerAwaiterTask.IsCompleted)
+                this.triggerAwaiterTask = this.trigger.WaitAsync(this.cancellation.Token);
+
+            List<Task> awaitedTasks = new List<Task>();
+            awaitedTasks.Add(this.triggerAwaiterTask);
+
+            if (this.timerAwaiterTask != null && !this.timerAwaiterTask.IsCompleted)
+                awaitedTasks.Add(this.timerAwaiterTask);
+
+            // Wait until trigger is set or timer task is completed.
+            await Task.WhenAny(awaitedTasks).ConfigureAwait(false);
+
+            // Reset trigger.
+            if (this.trigger.IsSet)
+                this.trigger.Reset();
+
+            // Process queue if task wasn't canceled.
+            if (awaitedTasks.Any(x => x.Status == TaskStatus.RanToCompletion))
+                this.processingAction(this.queue);
+
+            // Timer is no longer needed.
+            this.timerAwaiterTask = null;
+
+            if (this.restartTimerIfNotAllItemsConsumed &&
+                (this.timer != null) &&
+                (this.Count != 0) &&
+                !this.cancellation.IsCancellationRequested)
+            {
+                // Restart timer and recursively call execution flow.
```

Comment:
recursively!?

Reply:
Yes. 
I could have made a loop here but it looks like in most cases this loop will have only 1 iteration.